File: code/go/ch17/src/luago/compiler/compiler.go
package compiler

import "luago/binchunk"
import "luago/compiler/codegen"
import "luago/compiler/parser"

func Compile(chunk, chunkName string) *binchunk.Prototype {
	ast := parser.Parse(chunk, chunkName)
	return codegen.GenProto(ast)
}

File: code/java/ch18/src/main/java/com/github/zxh0/luago/compiler/Compiler.java
package com.github.zxh0.luago.compiler;

import com.github.zxh0.luago.binchunk.Prototype;
import com.github.zxh0.luago.compiler.ast.Block;
import com.github.zxh0.luago.compiler.codegen.CodeGen;
import com.github.zxh0.luago.compiler.parser.Parser;

public class Compiler {

    public static Prototype compile(String chunk, String chunkName) {
        Block ast = Parser.parse(chunk, chunkName);
        Prototype proto = CodeGen.genProto(ast);
        setSource(proto, chunkName);
        return proto;
    }

    private static void setSource(Prototype proto, String chunkName) {
        proto.setSource(chunkName);
        for (Prototype subProto : proto.getProtos()) {
            setSource(subProto, chunkName);
        }
    }

}
